### ISSUE TITLE
Audio / video capture is not working on macOS Catalina

### ISSUE THREAD

**Issue by arudnev** [nwjs/nw.js#7330]
NWJS Version : 0.43.6
Operating System : macOS Catalina (10.15.2)

### Expected behavior

navigator.mediaDevices.getUserMedia should return stream with audio and video tracks, permissions for app to use microphone / camera should be requested if needed

### Actual behavior

"DOMException: Could not start video source" is thrown or stream is returned with audio track that can not actually use microphone

### How to reproduce

run either of the following in dev tools

```javascript
navigator.webkitGetUserMedia({audio: true, video: true}, stream => console.info(stream), err => console.error(err))
```

```javascript
navigator.mediaDevices.getUserMedia({video:true, audio: true}).then(stream => console.info(stream)).catch(err => console.error(err))
```

We've noticed this issue on recent releases of macOS Catalina.
If permissions for use of microphone or camera were requested on macOS Mojave then everything works fine after upgrade to macOS Catalina, but if it's new install or permissions were reset via `tccutil reset Microphone` / `tccutil reset Camera` following instructions from https://developer.apple.com/documentation/avfoundation/cameras_and_media_capture/requesting_authorization_for_media_capture_on_macos then permissions are not requested and getUserMedia fails with `DOMException`.

Adding `videoCapture` / `audioCapture` permissions in manifest does not seem to change anything.

If `NSMicrophoneUsageDescription ` / `NSCameraUsageDescription` are added to `Info.plist` then exception is not thrown, but permissions are not requested and media tracks in resulting stream don't provide any audio / video, don't activate microphone / camera.
 
Is there something like `systemPreferences.askForMediaAccess(mediaType)` https://electronjs.org/docs/all#systempreferencesaskformediaaccessmediatype-macos that we are missing to call?

**Comment by arudnev** [comment]
It seems that adding appropriate combination of entitlements for App Sandbox or Hardened Runtime solves the issue

[com.apple.security.device.audio-input](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_device_audio-input)
[com.apple.security.device.microphone](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_device_microphone)
[com.apple.security.device.camera](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_device_camera)

Also, `NSMicrophoneUsageDescription` / `NSCameraUsageDescription` are set to something like "Once Chromium has access, websites will be able to ask you for access." in localized strings, so changing that is the only remaining cosmetic step.

**Comment by djipco** [comment]
Would you care to elaborate on what you did to solve this issue? I am having the same problem and I'm not sure how to solve it.

**Comment by arudnev** [comment]
I've added those 3 entitlements, so the file used in codesign process looks something like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>com.apple.security.device.camera</key>
    <true/>
    <key>com.apple.security.device.audio-input</key>
    <true/>
    <key>com.apple.security.device.microphone</key>
    <true/>
    <key>com.apple.security.automation.apple-events</key>
    <true/>
    <key>com.apple.security.cs.allow-dyld-environment-variables</key>
    <true/>
    <key>com.apple.security.cs.allow-jit</key>
    <true/>
    <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
    <true/>
    <key>com.apple.security.cs.disable-executable-page-protection</key>
    <true/>
    <key>com.apple.security.cs.disable-library-validation</key>
    <true/>
  </dict>
</plist>
```

I presume one would use some scripts for signing and notarization along the lines of what's described in #7117 

We use something like this for signing (I believe there is more generic code available in some of those tickets):

```bash
#!/bin/bash

export CODESIGN_ALLOCATE="/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate"

# Run the following to get a list of certs: security find-identity
identity="${1}"
# Location of the app
app="${2}"
# Path to the entitlements file
entitlements="${3:-neededToRun.entitlements}"

framework="Contents/Frameworks/nwjs Framework.framework"
base="${framework}/Versions/Current"

parts=(
  "${base}/libffmpeg.dylib"
  "${base}/libnode.dylib"
  "${base}/Libraries/libEGL.dylib"
  "${base}/Libraries/libGLESv2.dylib"
  "${base}/Libraries/libswiftshader_libEGL.dylib"
  "${base}/Libraries/libswiftshader_libGLESv2.dylib"
  "${base}/Helpers/chrome_crashpad_handler"
  "${base}/Helpers/app_mode_loader"
  "${base}/Helpers/nwjs Helper.app"
  "${base}/Helpers/nwjs Helper (GPU).app"
  "${base}/Helpers/nwjs Helper (Plugin).app"
  "${base}/Helpers/nwjs Helper (Renderer).app"
  "${base}/XPCServices/AlertNotificationService.xpc"
  # "${base}/nwjs Framework"
  # "${base}"
  "${framework}"
)

for part in "${parts[@]}"
do
  echo "### signing ${part}"
  codesign --force --verbose=4 --strict --options runtime --timestamp --sign "${identity}" --entitlements "${entitlements}" "${app}/${part}"
done

echo "### signing app"
codesign --force --verbose=4 --strict --options runtime --timestamp --deep --sign "${identity}" --entitlements "${entitlements}" "${app}"

echo "### verifying signature"
codesign --verify --verbose=4 "${app}"
spctl --assess --verbose=4 "${app}"
```

And then something similar to https://github.com/electron/electron-notarize for notarization.

We have not done update of localized strings that are used in permission prompt dialogs, but one would be looking for some process to update some of the values in InfoPlist.strings files as part of the app build process. See something like this for details:

`cat ./nwjs-sdk-v0.44.0-osx-x64/nwjs.app/Contents/Resources/*.lproj/InfoPlist.strings | grep UsageDescription`

**Comment by mikekilburn** [comment]
@arudnev What version of nwjs did you get this working? Our app is successfully notarized with hardened runtime and has the camera and mic entitlements yet Catalina will never prompt when call getUserMedia. This is 0.45.  Older version without the hardened runtime (based on 0.20) and no entitlements we get the OS prompts. It seems like this is an nwjs issue.

**Comment by arudnev** [comment]
Hm, strange, we fixed our app when it was still on 0.43.x, it has been working for us since then, including 0.45.5, which is the latest at the moment.

I remember that at some point we did run into some issues because permission request from our app was blocked by similar requests for zoom or skype that got queued and ignored by us as part of testing, but reboot or going into System Preference -> Security & Privacy brought all those queued pop-ups back to live.

I'll re-test with 0.45.5 once again just in case to make sure that we did not miss it somehow.
